draft-ietf-mls-protocol.md

```diff
@@ -1435,44 +1465,230 @@ all arrive at the following state:
+- The length of the key being generated.
 
-In combination with server-side enforced ordering, data from previous messages
-is used (as a salt when hashing) to:
+The traffic keying material is generated from an input traffic secret value using:
 
- * add freshness to derived symmetric keys
- * cryptographically bind the transcript of all previous messages with the current
-   group shared secret
+~~~~
+write_nonce_[sender] =
+  HKDF-Expand-Label(Application_Secret,"mls app nonce", [sender], nonce_length)
 
-Possible candidates for that are:
+write_key_[sender] =
+  HKDF-Expand-Label(Application_Secret,"mls app key", [sender], key_length)
+~~~~
 
- * the key used for the previous message (hash ratcheting)
- * the counter of the previous message (needs to be known to new members of the group)
- * the hash of the previous message (proof that other participants saw the same history)
+In this figure, [sender] represents the four-byte value representing the
+participant index in the tree used for the group key establishment mechanism.
+Note, that because the identity of the participant using the keys to send data
+is included in the initial Application Secret, all successive updates to the
+Application secret will implicitely inherit this ownership.
 
-The requirement for this is that all participants know these values.
-If additional clear-text fields are attached to messages (like the counter), those
-fields MUST be protected by a signed message envelope.
+All the traffic keying material is recomputed whenever the underlying
+Application Secret changes.
+
+
+## Message Encryption and Decryption
+
+The Group participants MUST use the AEAD algorithm associated with
+the negotiated MLS ciphersuite to AEAD encrypt and decrypt their
+Application messages and sign them as follows:
+
+~~~~~
+    struct {
+        uint32 length;
+        opaque content[length];
+        opaque signature<0..2^16-1>;
+        uint8 zeros[length_of_padding];
+    } MLSPlaintext;
+
+    struct {
+        uint8  group[32];
+        uint32 epoch;
+        uint32 generation;
+        uint32 sender;
+        uint32 length;
+        opaque encrypted_content<0..2^32-1>;
+    } MLSCiphertext;
+~~~~~
+
+The Group identifier and epoch allows a device to know which Group secret
+should be used and from which Epoch secret to start computing other secrets
+and keys. The participant identifier is used to derive the participant
+Application secret chain from the initial shared Application secret.
+The application generation field is used to determine which Application
+secret should be used from the chain to compute the correct AEAD keys
+before performing decryption.
+
+The signature field, allows to provide strong authentication of the
+plaintext and avoid Group participants to impersonate other participants.
+The signature MUST cover the metadata information about the current state
+of the group (group identifier, epoch, generation and sender) to avoid
+cross-group attacks.
```

Comment:
This is not a normative requirement, because the normative requirement is the structure below. Just move this to afterwards and then use it as an explanation for why.

Reply:
Reordered and made a minor rewrite to eliminate more grammatical mistakes.